File: internal/job/feed.go
package job

import (
	"cmp"
	"slices"
	"strings"

	"github.com/egladman/magus/internal/trail"
	"github.com/egladman/magus/types"
)

// Attribution is which job a changed path belongs to, and the declaration that decided it.
//
// Ambiguous is the failure mode spelled out rather than hidden in a bool: when two live
// boundaries cover one path there is no tie to break, and the reader needs the two names
// to go and fix the plan. See [AttributeWrite].
type Attribution struct {
	// Job is the one live job whose write paths cover the path, empty when none does or
	// more than one does.
	Job string
	// WritePath is the declared write path that covered it, so a reader sees WHY the path
	// was attributed and not merely that it was.
	WritePath string
	// Ambiguous names every live job that claimed the path, set only when more than one
	// did, sorted so the pair reads the same way twice.
	Ambiguous []string
}

// AttributeWrite names the one LIVE job whose declared write paths cover rel, a
// repository-relative path. It is what lets a person watch a worker without the worker
// cooperating: nothing in the write itself says who made it, and the declaration does.
//
// THE ATTRIBUTION IS EXACT BECAUSE THE WRITE PATHS ARE DISJOINT, and that is a property the
// fork already proves rather than one assumed here: [RefuseSharedCheckout] refuses a fork
// whose write paths would collide, [Store.WriteProof] records what it could prove, and `magus
// ls jobs` prints every pair that claims common ground. So the usual answer is one job or
// none.
//
// When two live jobs' write paths DO cover one path the answer is NEITHER, with both names
// in Ambiguous. Picking one would tell a person a file moved under a worker that never
// touched it, and there is nothing in a path to break the tie with; the plan is already
// damaged at that point, and saying so is more use than a coin toss. A job's own deny
// paths cut its boundary first, so a shared build input inside a directory somebody leased
// is nobody's.
//
// A job that has ended holds nothing. Rows are kept after they finish, and a finished
// boundary that kept answering would collect the next holder's writes for as long as the
// row lived.
func AttributeWrite(rows []types.Job, rel string) (Attribution, bool) {
	var out Attribution
	var claimants []string
	for _, row := range rows {
		if !row.State.Live() {
			continue
		}
		declared, covered := matching(row.WritePaths, rel)
		if !covered {
			continue
		}
		if _, denied := matching(row.DenyPaths, rel); denied {
			continue
		}
		claimants = append(claimants, row.ID)
		out.Job, out.WritePath = row.ID, declared
	}
	if len(claimants) > 1 {
		slices.Sort(claimants)
		return Attribution{Ambiguous: claimants}, false
	}
	return out, out.Job != ""
}

// Covers reports whether a declared path covers a repository-relative one, on the rule the
// write paths are graded by: a glob matches as a glob, and a literal directory covers what
// is under it.
//
// Exported so a reader FILTERING by path asks the same question the guard asks when it
// grades a write. Two answers to "is this file inside that path" is how a console filter
// comes to hide the write a deny was about.
func Covers(declared, p string) bool { return covers(declared, p) }

// Feed event kinds. They are the three things a watcher can see a worker do, and each one
// comes from a different producer: the filesystem, the guard, and the run journal.
const (
	FeedFile = "file" // a path under the job's write paths changed
	FeedTool = "tool" // the guard observed a tool call under the job's lease
	FeedRun  = "run"  // a run magus recorded against the job
)

// FeedEvent is one line of a job's feed: something that happened, when, and how it was
// attributed. It is deliberately flat and small. This rides a stream that a terminal
// prints one line per event from and a drawer appends a row per event to, so a shape
// either of them has to unpack first would be paid for on every event.
type FeedEvent struct {
	Ts   int64  // unix milliseconds
	Kind string // one of FeedFile, FeedTool, FeedRun
	Job  string // the job this was attributed to, empty when nothing could attribute it
	// WritePath is the declaration that attributed a file event. Empty on the other kinds,
	// which are attributed by lease rather than by path.
	WritePath string
	// Action is what happened, in the kind's own vocabulary: the path for a file event,
	// the tool label for a tool call, the rendered command for a run.
	Action string
	// Actor is the recorded origin rendered as one label (types.Origin.Label).
	Actor   string
	Host    string
	Session string
	// Decision is the guard's verdict on a tool call: pass, advise, ask or deny. Empty on
	// an observation the guard did not judge, which is what --observe records, and on the
	// other two kinds.
	Decision string
	// Outcome is trail.OutcomeOK or trail.OutcomeError, and Error carries the text.
	Outcome string
	Error   string
	// Ref is the output ref behind a run, so a reader can open the log the feed line names
	// rather than going to find it.
	Ref string
	// Note is the one thing about this line the fields above cannot say: the gate a run
	// belongs to, or that a path is contested.
	Note string
	// Contested names every live job that claimed this path, set only when more than one
	// did, and Job is then empty. It is a LIST rather than a sentence because a reader
	// watching one job has to ask "am I one of these", and parsing prose to answer that is
	// how the one event a contested plan most needs to surface gets dropped.
	//
	// Not hypothetical: measured in this repository, two live jobs both declared
	// internal/trail, and a watcher filtering on Job alone saw nothing at all while that
	// directory was being edited.
	Contested []string
}

// FileEvents attributes changed paths to live jobs, newest-first order preserved from the
// caller's batch. An unattributable path still produces an event, with an empty Job: a
// write outside every declared boundary is the one a reader most wants to see, and dropping
// it is how a view comes to show a quiet tree while somebody is editing it.
func FileEvents(rows []types.Job, ts int64, paths []string) []FeedEvent {
	out := make([]FeedEvent, 0, len(paths))
	for _, p := range paths {
		at, _ := AttributeWrite(rows, p)
		e := FeedEvent{Ts: ts, Kind: FeedFile, Job: at.Job, WritePath: at.WritePath, Action: p, Outcome: trail.OutcomeOK}
		if len(at.Ambiguous) > 0 {
			e.Contested = at.Ambiguous
			e.Note = "contested: " + strings.Join(at.Ambiguous, " and ") + " both declare this path"
		}
		out = append(out, e)
	}
	return out
}

// ToolEvents projects the guard's trail onto the feed, attributed BY LEASE.
//
// By lease and not by session, though the event carries both. One session declares many
// jobs and an orchestrator's session declares every job in the plan, so a session id names
// no single job and joining on it would attribute the orchestrator's own commands to
// whichever row it happened to have written last. A missing join is a gap a reader can
// see; a wrong one is a gap they cannot. Session stays on the event as something to FILTER
// by, which is a question the caller asked rather than a claim magus made.
func ToolEvents(events []trail.Event) []FeedEvent {
	out := make([]FeedEvent, 0, len(events))
	for _, e := range events {
		if e.Kind != trail.KindAgentCommand && e.Kind != trail.KindSandboxDenial {
			continue
		}
		out = append(out, FeedEvent{
			Ts:       e.Ts,
			Kind:     FeedTool,
			Job:      e.Lease,
			Action:   e.Action,
			Actor:    e.Label(),
			Host:     e.Host,
			Session:  e.Session,
			Decision: decisionOf(e),
			Outcome:  e.Outcome,
			Error:    e.Error,
		})
	}
	return out
}

// decisionOf reads the guard's verdict off the event without opening its payload blob. The
// preview is written as "guard: <decision>" by the producer, and "observed" when the guard
// judged nothing; a feed line that had to fetch a blob to say "deny" would cost one round
// trip per row.
func decisionOf(e trail.Event) string {
	const prefix = "guard: "
	if rest, ok := strings.CutPrefix(e.Preview, prefix); ok {
		return rest
	}
	return ""
}

// RunEvents is every run magus recorded against one job: the primary check, each
// completion gate, and the daemon's own last run for a catalog job. They are facts the
// store already holds, so the feed serves them without asking the output store anything.
func RunEvents(row types.Job) []FeedEvent {
	out := make([]FeedEvent, 0, 2+len(row.GateAttempts))
	if row.Attempt != nil && row.Attempt.Found {
		out = append(out, runEvent(row.ID, *row.Attempt, "check"))
	}
	for _, gate := range row.GateAttempts {
		if gate.Attempt.Found {
			out = append(out, runEvent(row.ID, gate.Attempt, "gate "+gate.GateID))
		}
	}
	if run := row.LastRun; run != nil && run.Ended > 0 {
		e := FeedEvent{Ts: run.Ended, Kind: FeedRun, Job: row.ID, Action: run.Invocation, Ref: run.Invocation, Outcome: trail.OutcomeOK, Note: "last run"}
		if !run.OK {
			e.Outcome, e.Error = trail.OutcomeError, run.Error
		}
		out = append(out, e)
	}
	return out
}

func runEvent(id string, a types.JobAttempt, note string) FeedEvent {
	e := FeedEvent{Ts: a.TimestampMs, Kind: FeedRun, Job: id, Action: a.String(), Ref: a.Ref, Outcome: trail.OutcomeOK, Note: note}
	if a.Failed {
		e.Outcome = trail.OutcomeError
	}
	return e
}

// FeedFilter narrows a feed to what one reader asked for. The three fields AND together
// and an empty one does not constrain, matching the ActivityQuery grammar the console
// already speaks.
//
// Paths match the same way a declared write path does, so asking for a directory answers
// for what is under it.
type FeedFilter struct {
	Jobs     []string
	Sessions []string
	Paths    []string
}

// Match reports whether e is what the filter asked for.
//
// A job filter matches a CONTESTED event that names it, even though the event is attributed
// to nobody. The two facts are different questions: "whose write is this" has no answer, and
// "is this something the person watching that job needs to see" plainly does.
func (f FeedFilter) Match(e FeedEvent) bool {
	if len(f.Jobs) > 0 && !slices.Contains(f.Jobs, e.Job) && !slices.ContainsFunc(f.Jobs, func(id string) bool {
		return slices.Contains(e.Contested, id)
	}) {
		return false
	}
	if len(f.Sessions) > 0 && !slices.Contains(f.Sessions, e.Session) {
		return false
	}
	if len(f.Paths) > 0 {
		if e.Kind != FeedFile {
			return false
		}
		if _, ok := matching(f.Paths, e.Action); !ok {
			return false
		}
	}
	return true
}

// Empty reports whether the filter constrains nothing, which is what lets a caller skip
// the walk entirely.
func (f FeedFilter) Empty() bool {
	return len(f.Jobs) == 0 && len(f.Sessions) == 0 && len(f.Paths) == 0
}

// SortFeed orders events newest first, stably, so events sharing a millisecond keep the
// order their producer emitted them in rather than shuffling between calls.
func SortFeed(events []FeedEvent) {
	slices.SortStableFunc(events, func(a, b FeedEvent) int { return cmp.Compare(b.Ts, a.Ts) })
}

// Ascending reverses a newest-first trail read. A follower reads forward, and reversing at
// the seam keeps the one sort in the reader rather than growing a second policy about what
// "in order" means.
func Ascending(events []trail.Event) []trail.Event {
	out := slices.Clone(events)
	slices.Reverse(out)
	return out
}

// TrailCursor is how far forward a follower has read the trail.
//
// It carries a COUNT alongside the timestamp because the trail's resolution is a
// millisecond and two observations routinely share one: a cursor holding only a timestamp
// either replays that millisecond's events on every tick or drops the second of them, and
// both read to a person as a bug in the guard.
//
// Shared by the streaming RPC and `magus job watch` so the two cannot disagree about what
// "since I last looked" means.
type TrailCursor struct {
	ts   int64
	seen int
}

// CursorAt places a cursor at the end of asc, so a follower that skipped the backfill does
// not receive the whole retained window on its first tick.
func CursorAt(asc []trail.Event) *TrailCursor {
	c := &TrailCursor{}
	if len(asc) == 0 {
		return c
	}
	c.ts = asc[len(asc)-1].Ts
	c.seen = countAt(asc, c.ts)
	return c
}

// Next returns the events in asc this cursor has not yielded, and advances past them.
func (c *TrailCursor) Next(asc []trail.Event) []trail.Event {
	var out []trail.Event
	at := 0
	for _, e := range asc {
		if e.Ts < c.ts {
			continue
		}
		if e.Ts == c.ts {
			at++
			if at <= c.seen {
				continue
			}
		}
		out = append(out, e)
	}
	if len(out) == 0 {
		return nil
	}
	c.ts = out[len(out)-1].Ts
	c.seen = countAt(asc, c.ts)
	return out
}

func countAt(asc []trail.Event, ts int64) int {
	n := 0
	for _, e := range asc {
		if e.Ts == ts {
			n++
		}
	}
	return n
}

// RunCursor remembers which of a job's recorded runs have been reported.
//
// A run is identified by its job, its output ref and the note naming which check it was,
// because a row's runs are REWRITTEN IN PLACE rather than appended: a re-run keeps its gate
// id and takes a new ref, so a cursor keyed on time alone would either re-report the whole
// set every tick or miss the re-run entirely.
type RunCursor struct{ sent map[string]bool }

func NewRunCursor() *RunCursor { return &RunCursor{sent: map[string]bool{}} }

// Prime marks everything the store already holds as reported, without yielding it.
func (c *RunCursor) Prime(rows []types.Job) {
	for _, row := range rows {
		for _, e := range RunEvents(row) {
			c.sent[runKey(e)] = true
		}
	}
}

// Next returns the runs recorded since the last call.
func (c *RunCursor) Next(rows []types.Job) []FeedEvent {
	var out []FeedEvent
	for _, row := range rows {
		for _, e := range RunEvents(row) {
			key := runKey(e)
			if c.sent[key] {
				continue
			}
			c.sent[key] = true
			out = append(out, e)
		}
	}
	return out
}

func runKey(e FeedEvent) string { return e.Job + "\x00" + e.Ref + "\x00" + e.Note }

File: console/src/console/plan/feed.ts
// The Jobs drawer's live feed: what the selected job's holder is DOING, right now, without
// anybody asking it.
//
// Everything else in this drawer is a declaration or a verdict - what a job was handed, what
// it has released, whether its gates passed. Those answer "is it done". This answers "is it
// going", which is the question somebody actually has at 4pm when a worker has been quiet
// for twenty minutes and interrupting it would cost it the turn it is in.
//
// THREE PRODUCERS, ONE SUBSCRIPTION. The server merges file changes attributed by write
// lane, the guard's tool-call observations attributed by lease, and the runs recorded
// against the job, and it time-orders them before they reach here (see
// magus.activity.v1alpha1.ActivityService.WatchActivityEvents). A client stitching three
// feeds would show a deny after the run it blocked.

import { createClient } from "@connectrpc/connect";
import { ActivityService, Kind, Outcome } from "@wire/activity/v1alpha1/activity_pb";
import type { ActivityEvent } from "@wire/activity/v1alpha1/activity_pb";
import { createServerTransport, getLiveToken } from "../../lib/server";
import { h } from "../view";

// BACKFILL is what the drawer asks for when it opens. A stream that started at "now" would
// paint a blank panel onto a job that has been running for an hour, and a blank panel reads
// as "nothing has happened" rather than as "I only just started listening".
const BACKFILL = 60;

// KEEP bounds the rows held for one job. A worker editing a directory can produce hundreds of
// changes a minute, and a drawer is read by scrolling the last screenful: past this the
// oldest rows are dropped rather than grown into a session-long transcript nobody reads.
const KEEP = 200;

// FeedRow is one rendered line. Flat and pre-formatted, because the render path appends one
// of these per event and a shape it had to interpret would be interpreted on every one.
export interface FeedRow {
  readonly at: number; // unix milliseconds
  readonly kind: "file" | "tool" | "run" | "other";
  readonly label: string; // the path, the tool, or the command
  readonly mark: string; // the guard's verdict, or a run's outcome
  readonly note: string; // contested paths, which gate a run was, why it failed
  readonly bad: boolean; // a deny, an ask, or a failed run: the rows somebody is scanning for
}

// rowOf projects one wire event onto a line. An event whose kind this console does not know
// still renders, as "other" with its action: a feed that silently dropped what it could not
// classify would go quiet exactly when the server grew something new to say.
export function rowOf(e: ActivityEvent): FeedRow {
  const at = e.time ? Number(e.time.seconds) * 1000 + Math.floor(e.time.nanos / 1e6) : 0;
  const failed = e.outcome === Outcome.ERROR;
  switch (e.kind) {
    case Kind.FILE_CHANGE:
      return {
        at,
        kind: "file",
        label: e.action,
        // A contested path is attributed to NOBODY, and both claimants are told. The mark
        // says so rather than leaving the row looking like an ordinary edit, because the
        // fact worth seeing is that the plan has two live lanes over one file.
        mark: e.contested.length ? "contested" : "",
        note: e.contested.length
          ? e.contested.join(" and ") + " both declare this path"
          : e.preview,
        bad: e.contested.length > 0,
      };
    case Kind.RUN:
      return {
        at,
        kind: "run",
        label: e.action,
        mark: failed ? "failed" : "ok",
        note: [e.preview, e.error].filter(Boolean).join(": "),
        bad: failed,
      };
    case Kind.AGENT_COMMAND:
    case Kind.SANDBOX_DENIAL: {
      // The guard writes its verdict into the preview as "guard: <decision>", and "observed"
      // when it judged nothing. Read off the event rather than fetched, so a two-hundred-row
      // feed does not cost two hundred payload reads to say "deny".
      const verdict = e.preview.startsWith("guard: ") ? e.preview.slice(7) : "";
      return {
        at,
        kind: "tool",
        label: e.action,
        mark: verdict,
        note: e.error,
        bad: verdict === "deny" || verdict === "ask" || failed,
      };
    }
    default:
      return { at, kind: "other", label: e.action, mark: "", note: e.preview, bad: failed };
  }
}

// JobFeed holds one job's rows and the element showing them. It owns its own element so the
// detail sheet can be rebuilt around it without the feed losing its place: a stream that
// restarted every time a poll tick repainted the sheet would show the last second of a
// worker's life and nothing else.
export class JobFeed {
  readonly el: HTMLElement;
  private rows: FeedRow[] = [];
  private job = "";
  private abort: AbortController | null = null;
  private trouble = "";

  constructor() {
    this.el = h("div", "console-plan-detail__feed");
    this.render();
  }

  // follow points the feed at a job, or at nothing. Re-pointing it at the job it is already
  // following is a no-op rather than a restart, because the caller is a selection sync that
  // runs on every repaint.
  follow(host: string, job: string | null): void {
    if (job === (this.job || null)) return;
    this.stop();
    this.job = job ?? "";
    this.rows = [];
    this.trouble = "";
    this.render();
    if (!job || !host) return;
    this.abort = new AbortController();
    void this.stream(host, job, this.abort.signal);
  }

  // stop ends the subscription. Called when the selection moves and when the surface is
  // unmounted: a stream nobody is reading is a stream the server is still writing to.
  stop(): void {
    this.abort?.abort();
    this.abort = null;
  }

  private async stream(host: string, job: string, signal: AbortSignal): Promise<void> {
    try {
      const client = createClient(ActivityService, createServerTransport(host, getLiveToken()));
      const stream = client.watchActivityEvents(
        { backfill: BACKFILL, filter: { units: [job] } },
        { signal },
      );
      for await (const e of stream) {
        if (this.job !== job) return; // the selection moved while this was in flight
        this.push(rowOf(e));
      }
      // The stream ended without an error: the server closed it, or the page is going away.
      // Said plainly rather than left looking live, because a feed that stopped and still
      // looks like a feed is how a person concludes a worker went quiet.
      if (this.job === job) {
        this.trouble = "The feed ended. Reopen this job to start it again.";
        this.render();
      }
    } catch (err) {
      if (signal.aborted || this.job !== job) return;
      this.trouble = "The feed could not be read (" + why(err) + ").";
      this.render();
    }
  }

  private push(row: FeedRow): void {
    this.rows.push(row);
    if (this.rows.length > KEEP) this.rows = this.rows.slice(-KEEP);
    this.render();
  }

  private render(): void {
    const box = h("div", "console-plan-feed");
    const head = h("h3", "console-plan-detail__runshead", "Live");
    box.append(head);
    if (!this.job) {
      box.append(h("p", "console-plan-detail__hint", "Select a job to watch what it is doing."));
    } else if (this.trouble) {
      box.append(h("p", "console-plan-detail__hint", this.trouble));
    } else if (!this.rows.length) {
      box.append(
        h(
          "p",
          "console-plan-detail__hint",
          "Nothing yet. Files changed under this job's write paths, the tool calls the guard" +
            " saw under its lease, and its runs all appear here as they happen.",
        ),
      );
    } else {
      const ul = h("ul", "console-plan-feed__list");
      ul.setAttribute("role", "list");
      // Newest last, and the container scrolls: this reads like a terminal, which is what a
      // person watching work happen already knows how to read.
      for (const row of this.rows) ul.append(lineOf(row));
      box.append(ul);
    }
    this.el.replaceChildren(box);
    // Only when there is something to scroll to. Calling this on an empty list moves nothing
    // and still fights a reader who has scrolled up to read an older deny.
    const list = this.el.querySelector(".console-plan-feed__list");
    if (list) list.scrollTop = list.scrollHeight;
  }
}

function lineOf(row: FeedRow): HTMLElement {
  const li = h("li", "console-plan-feed__row");
  li.dataset.kind = row.kind;
  if (row.bad) li.dataset.bad = "";
  li.append(h("span", "console-plan-feed__time", clock(row.at)));
  li.append(h("span", "console-plan-feed__kind", row.kind));
  if (row.mark) li.append(h("span", "console-plan-feed__mark", row.mark));
  li.append(h("code", "console-plan-feed__label", row.label));
  if (row.note) li.append(h("span", "console-plan-feed__note", row.note));
  return li;
}

// clock is the wall time, not an age. An age on every row would have to tick, and a feed
// that rewrote every line once a second would fight the reader's scroll position.
function clock(ms: number): string {
  if (!ms) return "";
  return new Date(ms).toLocaleTimeString([], { hour12: false });
}

function why(e: unknown): string {
  return e instanceof Error ? e.message : String(e);
}
